### DISCUSSION TITLE
fatal error: sys/capability.h: No such file or directory (zapret v68)

### DISCUSSION THREAD

**Discussion by kavardax** [bol-van/zapret#794] [Q&A] [answered]
Доброго всем времени суток! Использую Linux Mint 21.2 Victoria.
Делаю так:
/opt/zapret$ sudo git pull
remote: Enumerating objects: 100, done.
remote: Counting objects: 100% (100/100), done.
remote: Compressing objects: 100% (55/55), done.
remote: Total 100 (delta 54), reused 90 (delta 45), pack-reused 0 (from 0)
Receiving objects: 100% (100/100), 191.32 KiB | 747.00 KiB/s, done.
Resolving deltas: 100% (54/54), completed with 11 local objects.
From https://github.com/bol-van/zapret
   5d6c91f7..53546a8d  master     -> origin/master
Updating 5d6c91f7..53546a8d
Fast-forward
 Makefile                                           |  13 +++
 blockcheck.sh                                      | 321 +++++++++++++++++++++++++++++++++++++++++----------------------------
 common/base.sh                                     |  13 ++-
 common/installer.sh                                |   4 +-
 docs/compile/build_howto_openwrt.txt               |  38 ++++++---
 docs/readme.md                                     |  12 ++-
 docs/wireguard/010-wg-mod.patch                    | 133 -----------------------------
 docs/wireguard/wireguard-mod.txt                   | 250 -----------------------------------------------------
 docs/{wireguard => }/wireguard_iproute_openwrt.txt |   2 +-
 ip2net/Makefile                                    |   4 +-
 mdig/Makefile                                      |   6 +-
 nfq/BSDmakefile                                    |   2 +-
 nfq/Makefile                                       |   4 +-
 nfq/darkmagic.c                                    |  22 +++--
 nfq/darkmagic.h                                    |  11 ++-
 nfq/desync.c                                       | 170 +++++++++++++++++++++++--------------
 nfq/dvtws                                          | Bin 0 -> 157392 bytes
 nfq/nfqws.c                                        |  25 ++++--
 nfq/params.h                                       |   4 +-
 tpws/BSDmakefile                                   |   2 +-
 tpws/Makefile                                      |   7 +-
 tpws/andr/_musl_license.txt                        |  26 ++++++
 tpws/andr/getifaddrs.c                             | 216 ++++++++++++++++++++++++++++++++++++++++++++++
 tpws/andr/ifaddrs.h                                |   8 ++
 tpws/andr/netlink.c                                |  54 ++++++++++++
 tpws/andr/netlink.h                                |  94 ++++++++++++++++++++
 tpws/helpers.c                                     |   7 +-
 tpws/params.h                                      |   4 +-
 tpws/tpws.c                                        |  19 +++--
 29 files changed, 839 insertions(+), 632 deletions(-)
 delete mode 100644 docs/wireguard/010-wg-mod.patch
 delete mode 100644 docs/wireguard/wireguard-mod.txt
 rename docs/{wireguard => }/wireguard_iproute_openwrt.txt (98%)
 create mode 100644 nfq/dvtws
 create mode 100644 tpws/andr/_musl_license.txt
 create mode 100644 tpws/andr/getifaddrs.c
 create mode 100644 tpws/andr/ifaddrs.h
 create mode 100644 tpws/andr/netlink.c
 create mode 100644 tpws/andr/netlink.h

Потом:
/opt/zapret$ sudo ./install_easy.sh 
* checking system
system is based on systemd
* checking executables
trying to compile
make: Entering directory '/opt/zapret'
make[1]: Entering directory '/opt/zapret/nfq'
rm -f nfqws dvtws winws.exe
make[1]: Leaving directory '/opt/zapret/nfq'
make[1]: Entering directory '/opt/zapret/tpws'
rm -f tpws *.o
make[1]: Leaving directory '/opt/zapret/tpws'
make[1]: Entering directory '/opt/zapret/ip2net'
rm -f ip2net *.o
make[1]: Leaving directory '/opt/zapret/ip2net'
make[1]: Entering directory '/opt/zapret/mdig'
rm -f mdig *.o
make[1]: Leaving directory '/opt/zapret/mdig'
make[1]: Entering directory '/opt/zapret/nfq'
cc -s -std=gnu99 -Os -o nfqws *.c crypto/*.c  -lnetfilter_queue -lnfnetlink -lz
In file included from nfqws.c:4:
**sec.h:9:10: fatal error: sys/capability.h: No such file or directory
    9 | #include <sys/capability.h>
      |          ^~~~~~~~~~~~~~~~~~
compilation terminated.**
In file included from sec.c:5:
**sec.h:9:10: fatal error: sys/capability.h: No such file or directory
    9 | #include <sys/capability.h>
      |          ^~~~~~~~~~~~~~~~~~
compilation terminated.**
make[1]: *** [Makefile:18: nfqws] Error 1
make[1]: Leaving directory '/opt/zapret/nfq'
make: *** [Makefile:6: all] Error 2
make: Leaving directory '/opt/zapret'
could not compile
make: Entering directory '/opt/zapret'
make[1]: Entering directory '/opt/zapret/nfq'
rm -f nfqws dvtws winws.exe
make[1]: Leaving directory '/opt/zapret/nfq'
make[1]: Entering directory '/opt/zapret/tpws'
rm -f tpws *.o
make[1]: Leaving directory '/opt/zapret/tpws'
make[1]: Entering directory '/opt/zapret/ip2net'
rm -f ip2net *.o
make[1]: Leaving directory '/opt/zapret/ip2net'
make[1]: Entering directory '/opt/zapret/mdig'
rm -f mdig *.o
make[1]: Leaving directory '/opt/zapret/mdig'
make: Leaving directory '/opt/zapret'

press enter to continue

И вот это:
/opt/zapret$ ./install_bin.sh 
no compatible binaries found

Что я делаю не так? Всем сапсибо!

**Comment by bol-van** [comment]
Или берем релиз с бинариками, или надо установить несколько dev пакетов для сборки

**Reply by kavardax** [reply]
Я прошу прощения, если брать релиз с бинарниками, то алгоритм обновления такой:
1. удалить старое
2. установить новое
?

**Reply by bol-van** [reply]
install_easy сам снесет старое, сохранив настройки

**Comment by 1nt** [comment]
Я установил эти и скомпилировалось.
sudo apt-get install libcap-dev libnetfilter-queue-dev

**Reply by bol-van** [reply, marked as answer]
Они в дистрах по-разному называются, но по сути нужен libcap (только хедеры), zlib и netfilter-queue

на дебиках
apt install make gcc zlib1g-dev libcap-dev libnetfilter-queue-dev

**Comment by ghost** [comment]
На Кинетике стоит вот это: https://bin.entware.net/mipssf-k3.4/

Там всего 4 библиотеки:
libcap-bin_2.69-1_mips-3.4.ipk
libcap-ng-bin_0.8.4-1_mips-3.4.ipk
libcap-ng_0.8.4-1_mips-3.4.ipk
libcap_2.69-1_mips-3.4.ipk

Установил их все, но всё равно ошибка:
make[1]: Entering directory '/opt/root/git/zapret/nfq'
cc -s -march=native -std=gnu99 -Os -flto=auto -o nfqws .c crypto/.c -lnetfilter_queue -lnfnetlink -lz
In file included from nfqws.c:4:
sec.h:9:10: fatal error: sys/capability.h: No such file or directory
#include <sys/capability.h>
^~~~~~~~~~~~~~~~~~
compilation terminated.
In file included from sec.c:5:
sec.h:9:10: fatal error: sys/capability.h: No such file or directory
#include <sys/capability.h>
^~~~~~~~~~~~~~~~~~
compilation terminated.



Есть идеи, чего еще не хватает?



P.S. Это тоже установлено:
zlib-dev_1.3.1-1_mips-3.4
zlib_1.3.1-1_mips-3.4
libnetfilter-queue_1.0.5-4_mips-3.4

**Reply by bol-van** [reply]
Не надо собирать на кинетике. Есть релиз с готовыми бинарями

**Comment by vehametra** [comment]
> sec.h:9:10: fatal error: sys/capability.h: Нет такого файла или каталога
>     9 | #include <sys/capability.h>
>       |          ^~~~~~~~~~~~~~~~~~
> compilation terminated.
> In file included from params.h:9,
>                  from ipset.h:5,
>                  from ipset.c:2:
> sec.h:9:10: fatal error: sys/capability.h: Нет такого файла или каталога
>     9 | #include <sys/capability.h>
>       |          ^~~~~~~~~~~~~~~~~~
> compilation terminated.
> In file included from params.h:9,
>                  from lua.c:42:
> sec.h:9:10: fatal error: sys/capability.h: Нет такого файла или каталога
>     9 | #include <sys/capability.h>
>       |          ^~~~~~~~~~~~~~~~~~
> compilation terminated.
> In file included from nfqws.c:4:
> sec.h:9:10: fatal error: sys/capability.h: Нет такого файла или каталога
>     9 | #include <sys/capability.h>
>       |          ^~~~~~~~~~~~~~~~~~
> compilation terminated.
> In file included from params.h:9,
>                  from params.c:1:
> sec.h:9:10: fatal error: sys/capability.h: Нет такого файла или каталога
>     9 | #include <sys/capability.h>
>       |          ^~~~~~~~~~~~~~~~~~
> compilation terminated.
> In file included from params.h:9,
>                  from protocol.c:5:
> sec.h:9:10: fatal error: sys/capability.h: Нет такого файла или каталога
>     9 | #include <sys/capability.h>
>       |          ^~~~~~~~~~~~~~~~~~
> compilation terminated.
> In file included from sec.c:5:
> sec.h:9:10: fatal error: sys/capability.h: Нет такого файла или каталога
>     9 | #include <sys/capability.h>
>       |          ^~~~~~~~~~~~~~~~~~
> compilation terminated.

> 
> edata/srss/zapret2$ aptitude show libpcap-dev libnetfilter-queue-dev 
> Пакет: libpcap-dev                               
> Версия: 1.10.4-4.1ubuntu3
> Состояние: установлен
> Установлен автоматически: нет
> Мультиархитектура: такая же
> Приоритет: необязательный
> Раздел: libdevel
> Сопровождающий: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
> Архитектура: amd64
> Размер в распакованном виде: 14,3 k
> Зависит: libpcap0.8-dev
> Ломает: libpcap-dev:i386 (!= 1.10.4-4.1ubuntu3)
> Заменяет: libpcap-dev:i386 (< 1.10.4-4.1ubuntu3)
> Описание: библиотека разработчика для libpcap (переходный пакет).
>  Пустой пакет для облегчения обновлений, может быть безопасно удален.
> Домашняя страница: https://www.tcpdump.org/
> 
> 
> Пакет: libnetfilter-queue-dev
> Версия: 1.0.5-4build1
> Состояние: установлен
> Установлен автоматически: нет
> Мультиархитектура: такая же
> Приоритет: необязательный
> Раздел: universe/libdevel
> Сопровождающий: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
> Архитектура: amd64
> Размер в распакованном виде: 76,8 k
> Зависит: libnetfilter-queue1 (= 1.0.5-4build1), libnfnetlink-dev, pkg-config
> Предлагает: libnetfilter-queue-doc (= 1.0.5-4build1)
> Описание: Development files for libnetfilter-queue1
>  libnetfilter_queue is a userspace library providing an API to packets that have been queued by the kernel packet
>  filter. It is part of a system that deprecates the old ip_queue / libipq mechanism. 


Ubuntu 24.04 , таже проблема, что не так?

**Reply by bol-van** [reply]
Если не умеете решать такие "проблемы", не лезте в сборку
Релиз есть готовый

**Reply by vehametra** [reply]
```
 ~$ apt-file list libpcap-dev
libpcap-dev: /usr/share/doc/libpcap-dev/changelog.Debian.gz
libpcap-dev: /usr/share/doc/libpcap-dev/copyright


 ~$ apt-file search  sys/capability.h
dietlibc-dev: /usr/include/diet/sys/capability.h
libcap-dev: /usr/include/sys/capability.h
libklibc-dev: /usr/lib/klibc/include/sys/capability.
```
А. libcap  а не libpcap.  _capability_ ,а не _Packet Capture_ .

**Reply by bol-van** [reply]
Совершенно верно. Вещи совсем разные
